### DISCUSSION TITLE
Obtaining compressed/uncompressed bytes transferred from ytdl.getInfo requests

### DISCUSSION THREAD

**Discussion by gatecrasher777** [fent/node-ytdl-core#899] [Ideas]
I have a github project that uses ytdl-core when stream urls are ciphered. I would like it to return the bytes transferred.  Before I fall back onto ytdl-core I use something very vanilla that can handle lots of simultaneous requests like this:

```javascript
  let transferred = 0;
  let zip = null;
  let getOptions = {
	  headers: {
		  'cookie': this.cookie,
		  'accept-encoding': 'br'
	  },
	  acceptEncoding: { 
		  br: () => { zip = zlib.createBrotliDecompress(); return zip; }
	  }
  }
  let req = miniget(url,getOptions);
  req.on('data', (chunk) =>{
	transferred += chunk.length;
  });
  let body = await req.text();
  if (zip !== null) transfered = this.zip.bytesWritten;
```

Similarly, when urls are missing/ciphered, and I fall back onto ytdl-core I can still implement br compression:

```javascript
  ytdl.getInfo( url. {
    requestOptions: {
	    headers: {
		    'cookie': this.cookie,
		    'accept-encoding': 'br'
	    },
	    acceptEncoding: { 
		    br: () => zlib.createBrotliDecompress()
	    }
    }
    }).then(info => {
          ...
    }, err => {
          ...
  });
```
but extracting the actual bytes transferred does not seem reliable.  I can get the bytes transferred on the last request ytdl-core makes with 
```javascript
  br: () => { zip = zlib.createBrotliDecompress(); return zip; }
```
and then querying zip.bytesWritten but its not a useful number if several requests are pipelined in ytdl-core prior to the last request.  

I know the bytes transferred to get video info is not important in the typical ytdl-core use case, but to be able to reliably monitor overall bandwidth in my project would be a big plus.   

I also don't want to use a modified ytdl-core in my project because ytdl-core updates are both often and crucial for reliable download urls.

I envision/hope for something simple and non-invasive like this:

```javascript
  let transferred;
  ytdl.getInfo( url. {
      ...
  }).then(info => {  
      ...
      transferred = info.transferred;
  } ,err => {
      ...
      transferred = err.transferred;
  });
```
Perhaps if I work on a pull request that achieves this and doesn't break anything, it might be considered? 

Thanks for a great  project.

**Comment by gatecrasher777** [comment]
Sorry to be impatient. I played around some and got a fairly nice solution:

In my code:  

```javascript
      let transferred;
      let uncompressed = 0;
      let compressed = 0;
      let zip = null;
      ytdl.getInfo( url, {
            reqCallback : (req) => {
	            req.on('data',(chunk)=>{
		            uncompressed += chunk.length;
	            });
	            req.on('end',()=>{
		            if (zip != null) compressed += zip.bytesWritten;
		            zip = null;
	            });
            },
            requestOptions: {
	            headers: {
		            'cookie': this.cookie,
		            'accept-encoding': 'gzip, deflate, br'
	            },
	            acceptEncoding: { 
		            gzip: () => { zip = zlib.createGunzip(); return zip; },
		            deflate: () => { zip = zlib.createDeflate(); return zip; },
		            br: () => { zip = zlib.createBrotliDecompress(); return zip; }
	            }
            }
      }).then(info => {
            (compressed) ? transferred += compressed : transferred += uncompressed;
            ...
            console.log('uncompressed: '+uncompressed);
            console.log('compressed: '+compressed);
      }, err => {
            (compressed) ? transferred += compressed : transferred += uncompressed;
            ...
            console.log('uncompressed: '+uncompressed);
            console.log('compressed: '+compressed);
});
```

In the modified info.js, each miniget request (about 6 of them) would expose the request stream. For example:
```javascript
    let req = miniget(jsonUrl, reqOptions);
    options.reqCallback(req);
    let body = await req.text();
```
instead of 
```javascript
    let body = await miniget(jsonUrl, reqOptions).text();
```

Typical output from the above looks like this:
```
    uncompressed: 494961
    compressed: 72311
    uncompressed: 520688
    compressed: 75746
    uncompressed: 516199
    compressed: 73751
    uncompressed: 513536
    compressed: 73960
    uncompressed: 505390
    compressed: 73839
    uncompressed: 535005
    compressed: 77828
```

So essentially, I'm suggesting a reqCallback option.  Exposing the miniget requests like this might have other applications no-one has thought of yet.

**Comment by mh4ck** [comment]
That sounds very interesting and should be implemented by default (without a 'reqCallback') if it is like you say. 

```
uncompressed: 494961
compressed: 72311
```

Is it only the info request or even videos can be received compressed?

**Reply by gatecrasher777** [reply]
Video formats are already highly compressed, so the bandwidth saving is on the info and html requests which youtube will serve back compressed as long as your request has the 'accept-encoding': 'gzip, deflate, br' header.

**Comment by fent** [comment]
all miniget events are forwarded to the ytdl stream, that includes the `request` event

https://github.com/fent/node-miniget#event-request

**Reply by gatecrasher777** [reply]
That is true for ytdl which returns a stream, but is that also true for ytdl.getInfo which retums the info? 

For example on line 129 of info.js
```javascript
   return miniget(embedUrl, options.requestOptions).text();
```

How is this miniget request forwarded? On the surface, there doesn't seem to be any access into it.  Unless I'm missing something obvious, which is quite possible at my age.

**Comment by mh4ck** [comment]
If we can get the request stream on a getInfo request it is also helpful to stop/handle broken proxy connections. In my case it happens sometimes that the connection to the proxy hangs up and the request never finish/error or run into a timeout.

If we get the request stream we can manually stop the connection and try again with another proxy.
